renegade-crypto/src/hash/poseidon2.rs

```diff
@@ -0,0 +1,297 @@
+    ///     [1, 1, 3]
+    ///
+    /// This can be done efficiently by adding the sum to each element as in the
+    /// external MDS case but adding an additional copy of the final state
+    /// element
+    #[cfg_attr(feature = "inline", inline)]
+    fn internal_mds(&mut self) {
+        let sum = self
+            .state
+            .iter()
+            .fold(ScalarField::zero(), |acc, x| acc + x);
+
+        self.state[WIDTH - 1].double_in_place();
+        for state_elem in self.state.iter_mut() {
+            *state_elem += sum;
+        }
+    }
+
+    /// Apply the s-box to an element of the state
+    ///
+    /// We hardcode the sbox parameterization to alpha = 5
+    #[cfg_attr(feature = "inline", inline)]
+    fn apply_sbox(val: &mut ScalarField) {
+        let base_val = *val;
+
+        val.square_in_place();
+        val.square_in_place();
+        val.mul_assign(&base_val);
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use ark_ff::BigInt;
+    use itertools::Itertools;
+    use rand::{thread_rng, Rng};
+    use zkhash::{
+        fields::bn256::FpBN256,
+        poseidon2::{poseidon2::Poseidon2, poseidon2_instance_bn256::POSEIDON2_BN256_PARAMS},
+    };
+
+    use crate::hash::{ScalarField, RATE, WIDTH};
+
+    use super::Poseidon2Sponge;
+
+    /// Test the permutation against a known implementation
+    #[test]
+    fn test_permutation() {
+        // Sample random values to permute on
+        let mut rng = thread_rng();
+        let values: [ScalarField; WIDTH] = rng.gen();
+
+        let expected_hasher = Poseidon2::new(&POSEIDON2_BN256_PARAMS);
+        let mut hasher = Poseidon2Sponge::new();
+
+        // Run the expected implementation's permutation
+        // We map into `BigInt` first here because although the implementations are
+        // defined over the same field, the reference implementation redefines
+        // the Bn254 scalar field, meaning it has a different runtime type than
+        // our scalar field
+        let expected_res = expected_hasher
+            .permutation(
+                &values
+                    .iter()
+                    .cloned()
+                    .map(BigInt::from)
+                    .map(FpBN256::from)
+                    .collect_vec(),
+            )
+            .into_iter()
+            .map(BigInt::from)
+            .collect_vec();
+
+        // Run our permutation
+        hasher.state = values;
+        hasher.permute();
+
+        // Compare the results
+        let res = hasher.state.into_iter().map(BigInt::from).collect_vec();
+        assert_eq!(expected_res, res);
+    }
+
+    /// Tests the hash by mocking a sponge with the reference
+    #[test]
+    fn test_hash() {
+        let mut rng = thread_rng();
+        let values: [ScalarField; RATE] = rng.gen();
+
+        let expected_hasher = Poseidon2::new(&POSEIDON2_BN256_PARAMS);
+        let mut hasher = Poseidon2Sponge::new();
+
+        // Run the expected implementation
+        let perm_input = [
```

Comment:
nit: I'm assuming the Horizen implementation expects the first input element to be zero, or something to that effect? would be nice to call this out explicitly in the comment above

Reply:
Yeah good call. It's actually to mimic the capacity element of a sponge, their impl only defines a permutation so we have to get a little hacky